src/main/kotlin/org/rust/ide/annotator/cargoCheck/RsCargoCheckAnnotator.kt

```diff
@@ -18,52 +18,92 @@ import com.intellij.openapi.editor.Editor
 import com.intellij.openapi.editor.impl.TrailingSpacesStripper
 import com.intellij.openapi.fileEditor.FileDocumentManager
 import com.intellij.openapi.fileEditor.impl.FileDocumentManagerImpl
-import com.intellij.openapi.module.ModuleUtil
 import com.intellij.openapi.progress.PerformInBackgroundOption
 import com.intellij.openapi.progress.ProgressIndicator
 import com.intellij.openapi.progress.ProgressManager
 import com.intellij.openapi.progress.impl.BackgroundableProcessIndicator
 import com.intellij.openapi.util.Computable
 import com.intellij.psi.PsiFile
+import com.intellij.psi.util.CachedValueProvider
+import com.intellij.psi.util.CachedValuesManager
+import com.intellij.psi.util.PsiModificationTracker
+import org.rust.cargo.project.model.CargoProject
 import org.rust.cargo.project.settings.rustSettings
 import org.rust.cargo.project.settings.toolchain
 import org.rust.cargo.project.workspace.PackageOrigin
 import org.rust.ide.annotator.cargoCheck.RsCargoCheckFilteredMessage.Companion.filterMessage
 import org.rust.lang.core.psi.RsFile
-import org.rust.lang.core.psi.ext.cargoWorkspace
+import org.rust.lang.core.psi.ext.cargoProject
 import org.rust.lang.core.psi.ext.containingCargoPackage
+import org.rust.openapiext.checkReadAccessAllowed
+import org.rust.openapiext.checkReadAccessNotAllowed
 import org.rust.openapiext.isUnitTestMode
 import org.rust.openapiext.saveAllDocuments
 import java.lang.reflect.Field
 
-object RsCargoCheckAnnotator : ExternalAnnotator<RsCargoCheckAnnotationInfo, RsCargoCheckAnnotationResult>() {
+object RsCargoCheckAnnotator : ExternalAnnotator<Lazy<RsCargoCheckAnnotationResult?>, RsCargoCheckAnnotationResult>() {
     private val LOG: Logger = Logger.getInstance(RsCargoCheckAnnotator::class.java)
     const val TEST_MESSAGE: String = "CargoCheckAnnotation"
 
-    override fun collectInformation(file: PsiFile): RsCargoCheckAnnotationInfo? {
+    override fun collectInformation(file: PsiFile): Lazy<RsCargoCheckAnnotationResult?>? {
+        checkReadAccessAllowed()
         if (file !is RsFile) return null
         if (!file.project.rustSettings.useCargoCheckAnnotator) return null
-        val ws = file.cargoWorkspace ?: return null
-        val module = ModuleUtil.findModuleForFile(file.virtualFile, file.project) ?: return null
-        val toolchain = module.project.toolchain ?: return null
-        val cargoPackage = file.containingCargoPackage
-        if (cargoPackage?.origin != PackageOrigin.WORKSPACE) return null
-        return RsCargoCheckAnnotationInfo(toolchain, ws.contentRoot, module, cargoPackage)
+
+        val cargoProject = file.cargoProject ?: return null
+        if (cargoProject.project.toolchain == null) return null
+        if (cargoProject.workspace?.contentRoot == null) return null
+
+        return CachedValuesManager.getManager(file.project)
+            .getCachedValue(cargoProject) {
```

Comment:
I think, we should ask @Undin to take a look to this comment :)

Reply:
In general, it's ok - at least I've understood the idea)
@mchernyavsky Let's add note to `collectInformation` java doc that result can be outdated